### PR TITLE
Us 11 13 - Inventory System

### PR DESCRIPTION
**Pull request by Alexander-LF**

### CODE REVIEW

File: Project/Fall2020_CSC403_Project/Fall2020_CSC403_Project.csproj [hunk 1]
```diff
@@ -8,10 +8,11 @@
     <OutputType>WinExe</OutputType>
     <RootNamespace>Fall2020_CSC403_Project</RootNamespace>
     <AssemblyName>Fall2020_CSC403_Project</AssemblyName>
-    <TargetFrameworkVersion>v4.7.2</TargetFrameworkVersion>
+    <TargetFrameworkVersion>v4.6.1</TargetFrameworkVersion>
```
Comment (james-love) [inline Project/Fall2020_CSC403_Project/Fall2020_CSC403_Project.csproj]: What's the reason for downgrading the framework?

File: Project/Fall2020_CSC403_Project/Fall2020_CSC403_Project.csproj [hunk 2]
```diff
@@ -60,14 +67,17 @@
...
+    </EmbeddedResource>
     <EmbeddedResource Include="FrmBattle.resx">
       <DependentUpon>FrmBattle.cs</DependentUpon>
     </EmbeddedResource>
     <EmbeddedResource Include="FrmLevel.resx">
       <DependentUpon>FrmLevel.cs</DependentUpon>
     </EmbeddedResource>
     <EmbeddedResource Include="Properties\Resources.resx">
-      <Generator>ResXFileCodeGenerator</Generator>
+      <Generator>PublicResXFileCodeGenerator</Generator>
```
Comment (james-love) [inline Project/Fall2020_CSC403_Project/Fall2020_CSC403_Project.csproj]: is this related to this issue?
https://stackoverflow.com/questions/4274311/visual-studio-resx-file-default-internal-to-public

File: Project/Fall2020_CSC403_Project/FrmLevel.cs [hunk 3]
```diff
@@ -147,6 +156,17 @@ private void OnEnemyDeath(Enemy enemy)
         case Keys.Down:
           player.GoDown();
           break;
+        case Keys.I:
+            ShowInventory();
+            break;
+
+        case Keys.C:
+          player.GetInventory().AddToInventory(new MyGameLibrary.InventoryObjects.HealthPotion(picHealthPotion.Image));
```
Comment (james-love) [inline Project/Fall2020_CSC403_Project/FrmLevel.cs]: Instead of homemade images, consider using free to use artwork such as this https://opengameart.org/content/16-bit-rpg-potion-pack

File: Project/MyGameLibrary/Inventory.cs [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+        {
+            inventory = new InventoryObject[maxInventorySize];
+        }
+
+        // Inventory constructor with passed List of InventoryObjects
+        public Inventory(int maxInventorySize,List<InventoryObject> invObjs)
+        {
+            inventory = new InventoryObject[maxInventorySize];
+            this.maxInventorySize = maxInventorySize;
+            for (int i = 0; i < invObjs.Count && i < maxInventorySize; i++)
```
Comment (james-love) [inline Project/MyGameLibrary/Inventory.cs]: Using a foreach would look cleaner. This is true of the other constructor as well.

File: Project/MyGameLibrary/Inventory.cs [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+            this.maxInventorySize = maxInventorySize;
+            for (int i = 0; i < invObjs.Length && i < maxInventorySize; i++)
+            {
+                inventory[i] = invObjs[i];
+            }
+        }
+
+        public int GetMaxInventorySize()
+        {
+            return maxInventorySize;
```
Comment (james-love) [inline Project/MyGameLibrary/Inventory.cs]: Instead of adding public accessor methods for private variable, consider something like this

```
private int _maxSize = 8;
public int MaxSize { get => _maxSize; }
```

Also, since its inside the inventory class, the full name of MaxInventorySize could be shortened to just MaxSize.

File: Project/MyGameLibrary/Inventory.cs [hunk 6]
```diff
@@ -0,0 +1,218 @@
+﻿using Fall2020_CSC403_Project.code;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
Comment (james-love) [inline Project/MyGameLibrary/Inventory.cs]: Several of this using statements aren't actually used in this file and should be removed. This is also true of the other new files in this review, so those should be checked as well. In Visual Studio you can quickly tell which ones are not being used as they will be slightly translucent and when you hover over them with the mouse it will tell you that they are unused.

File: Project/MyGameLibrary/Inventory.cs [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+                }
+            }
+
+            // There are no empty spaces in the inventory.
+            return false;
+        }
+
+        // Returns the index of the given object type (matches the passed
+        // object's class) in the inventory.
+        private int ObjectTypeExistsInInventory(InventoryObject iObj)
```
Comment (james-love) [inline Project/MyGameLibrary/Inventory.cs]: The name indicates that this would be a boolean check, however from the code this is really a find index returning -1 when there is none found. Consider replacing the function with this one that accomplishes the same task:
`return Array.FindIndex<InventoryObject>(inventory, x => x != null && x.TypeOf(iObj));`

File: Project/MyGameLibrary/Inventory.cs [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+                    }
+                }
+            }
+            return -1;
+        }
+
+        // Returns object in inventory at index without removing it.
+        public InventoryObject GetInventoryObject(int index)
+        {
+            //Console.WriteLine("\t" + index.ToString());
```
Comment (james-love) [inline Project/MyGameLibrary/Inventory.cs]: Remove unused commented out code

File: Project/MyGameLibrary/Inventory.cs [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+            if (index < inventory.Count())
+            {
+                //Console.WriteLine("\t" + inventory[index].GetType());
+                return (inventory[index]);
+            }
+            else
+            {
+                return null;
+            }
```
Comment (james-love) [inline Project/MyGameLibrary/Inventory.cs:144]: Its a matter of preference here on if you want to change it but this could be like this instead
```
if (index < inventory.Count())
    return (inventory[index]);
else
    return null;
```

File: Project/MyGameLibrary/Inventory.cs [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+                }
+            }
+            return -1;
+        }
+
+        // Returns object in inventory at index without removing it.
+        public InventoryObject GetInventoryObject(int index)
+        {
+            //Console.WriteLine("\t" + index.ToString());
+            if (index < inventory.Count())
```
Comment (james-love) [inline Project/MyGameLibrary/Inventory.cs]: This handles if the user passes and index larger than the array, but what if they pass a negative index? Consider instead wrapping this in a try catch and returning null on exception, taking care of this and any other case of an invalid index being passed

File: Project/MyGameLibrary/Inventory.cs [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+            else
+            {
+                return null;
+            }
+        }
+
+        // Removes an object at index from the inventory
+        public void RemoveInventoryObject(int index)
+        {
+            if(index < maxInventorySize)
```
Comment (james-love) [inline Project/MyGameLibrary/Inventory.cs]: Similar to the other comment, this doesn't handle negative numbers. See other comment for proposed solution.

File: Project/MyGameLibrary/Inventory.cs [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+            {
+                inventory[index] = null;
+            }
+        }
+
+        // Removes the given object from the inventory
+        public void RemoveInventoryObject(InventoryObject iObj)
+        {
+            for(int i = 0; i < maxInventorySize; i++)
```
Comment (james-love) [inline Project/MyGameLibrary/Inventory.cs]: Consider using a foreach instead

File: Project/MyGameLibrary/Inventory.cs [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+                {
+                    inventory[i] = null;
+                }
+            }
+        }
+
+        public void UseInventoryObject(int index, Player player)
+        {
+            InventoryObject iObj = GetInventoryObject(index);
+            if(iObj == null)
```
Comment (james-love) [inline Project/MyGameLibrary/Inventory.cs:171]: In general, there are several places with if statements that could have the braces removed, look over all the new code and determine which ones to remove.

File: Project/MyGameLibrary/Inventory.cs [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+            {
+                return;
+            }
+            iObj.Effect(player);
+            if (iObj.IsExhaustible())
```
Comment (james-love) [inline Project/MyGameLibrary/Inventory.cs]: All these if statements could be shortened to
```
if (iObj.IsExhaustible() && (!iObj.IsStackable() || iObj.GetCount() <= 0)
    RemoveInventoryObject(index);
```

File: Project/MyGameLibrary/InventoryItems/Experience.cs [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+    {
+        private readonly int experienceAmount = 1;
+        
+        public Experience(Image img, int count)
+        {
+            this.stackable = true;
+            this.exhaustible = true;
+            this.img = img;
+            this.count = count;
+
```
Comment (james-love) [inline Project/MyGameLibrary/InventoryItems/Experience.cs]: Extra new line added here

File: Project/MyGameLibrary/InventoryItems/HealthPotion.cs [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+    public class HealthPotion : InventoryObject
+    {
+        private readonly int healAmount = 6;
+        
+        public HealthPotion(Image img)
+        {
+            this.stackable = false;
+            this.exhaustible = true;
+            this.img = img;
+
```
Comment (james-love) [inline Project/MyGameLibrary/InventoryItems/HealthPotion.cs]: Extra new line added here

File: Project/MyGameLibrary/InventoryItems/HealthPotion.cs [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+            this.stackable = false;
+            this.exhaustible = true;
+            this.img = img;
+            
+        }
+
+        public override void Effect(BattleCharacter character)
+        {
+            throw new NotImplementedException();
+
```
Comment (james-love) [inline Project/MyGameLibrary/InventoryItems/HealthPotion.cs]: Extra new line added here, there are several places in other files as well. After making the other changes from this review, run through the code and remove them

File: Project/MyGameLibrary/InventoryObject.cs [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+            {
+                return false;
+            }
+            count -= removeCount;
+            return true;
+            
+        }
+
+        // Returns if this object is stackable
+        public bool IsStackable()
```
Comment (james-love) [inline Project/MyGameLibrary/InventoryObject.cs]: Instead of creating methods like this to access these fields, it would make more sense for them to be treated as properties for the object. This statement applies to all of the IsWhatever variables in this class as well as to the get count. Consider this
```
protected bool _stackable; 
public bool Stackable { get => stackable; }
```
Now later on instead of addressing the values as 
```
iObj.IsStackable()
iObj.IsExhaustible()
iObj.GetCount()
```
you can instead just use
```
iObj.Stackable
iObj.Exhaustible
iObj.Count
```

File: Project/MyGameLibrary/MyGameLibrary.csproj [hunk 19]
```diff
@@ -9,9 +9,10 @@
     <AppDesignerFolder>Properties</AppDesignerFolder>
     <RootNamespace>MyGameLibrary</RootNamespace>
     <AssemblyName>MyGameLibrary</AssemblyName>
-    <TargetFrameworkVersion>v4.7.2</TargetFrameworkVersion>
+    <TargetFrameworkVersion>v4.6.1</TargetFrameworkVersion>
```
Comment (james-love) [inline Project/MyGameLibrary/MyGameLibrary.csproj]: Why has the framework version been downgraded?

File: Project/MyGameLibrary/Player.cs [hunk 20]
```diff
@@ -1,13 +1,22 @@
...
 namespace Fall2020_CSC403_Project.code {
   public class Player : BattleCharacter {
+
+    private Inventory inventory;
+
     public Player(Vector2 initPos, Collider collider) : base(initPos, collider) {
+            inventory = new Inventory();
+    }
 
+    public Inventory GetInventory()
```
Comment (james-love) [inline Project/MyGameLibrary/Player.cs:17]: Similar to the comment I made in the InventoryObject class, should just use a property here instead of a separate accessor function.

File: Project/Fall2020_CSC403_Project/FrmInventory.cs [hunk 21]
```diff
@@ -0,0 +1,167 @@
...
+using System.Data;
+using System.Drawing;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Windows.Forms;
+
+namespace Fall2020_CSC403_Project
+{
+    public partial class FrmInventory : Form
```
Comment (james-love) [inline Project/Fall2020_CSC403_Project/FrmInventory.cs:8]: The inventory window currently shows up with a title of "Form1" this can be changed in the designer by selecting the overall window/form and changing the "Text" property.

File: Project/Fall2020_CSC403_Project/FrmInventory.cs [hunk 22]
```diff
@@ -0,0 +1,167 @@
...
+
+        // Creates an instance of this class with a visible 
+        // interface for the inventory
+        private void Instance()
+        {
+            InitializeComponent();
+            inventorySlots = new Button[inventory.GetMaxInventorySize()];
+            InitInvSlots();
+            this.FormClosing += new FormClosingEventHandler(FrmIventory_FormClosing);
+            invUpdater = new Thread(ContinuousUpdater);
```
Comment (james-love) [inline Project/Fall2020_CSC403_Project/FrmInventory.cs]: There is a bug which I think may be caused by this  where if you open the inventory then close the main game window without closing inventory first, you can no longer rebuild the project as the files will be locked by the still running process in the background which has to be manually killed.

File: Project/Fall2020_CSC403_Project/FrmInventory.cs [hunk 23]
```diff
@@ -0,0 +1,167 @@
...
+                        // Allows the inventory updater thread to update the UI elements
+                        // in the inventory from outside of the main thread.
+                        try
+                        {
+                            this.Invoke(new MethodInvoker(delegate ()
+                            {
+                                inventorySlots[i].Text = inventory.GetInventoryObject(i).GetCount().ToString();
+                            }));
+                        }
+                        catch(Exception e)
```
Comment (james-love) [inline Project/Fall2020_CSC403_Project/FrmInventory.cs]: Should we be doing anything when we catch all these exceptions? If we're not specifying the exception on any of these or doing anything once it's caught, why not wrap the whole thing in a single try catch?

**Review by james-love — CHANGES_REQUESTED** [review]

File: Project/MyGameLibrary/Inventory.cs (hunk 13 above)
Comment (Alexander-LF) [reply Project/MyGameLibrary/Inventory.cs:171]: I'm going to keep with the Google C# style guide we were recommended, which specifies to always use braces when optional.

File: Project/MyGameLibrary/Player.cs (hunk 20 above)
Comment (Alexander-LF) [reply Project/MyGameLibrary/Player.cs:17]: This would cause a conflict where the class name is the same as the property.

File: Project/MyGameLibrary/MyGameLibrary.csproj (hunk 19 above)
Comment (Alexander-LF) [reply Project/MyGameLibrary/MyGameLibrary.csproj]: My mistake, this is from using an older version of VS on my machine. It shouldn't change any functionality, but I'll update this before I push my updated code.

File: Project/Fall2020_CSC403_Project/Fall2020_CSC403_Project.csproj (hunk 2 above)
Comment (Alexander-LF) [reply Project/Fall2020_CSC403_Project/Fall2020_CSC403_Project.csproj]: Forgot to roll back the changes to this when from when I tried accessing the inventory images from the file instead of the form.

**Review by james-love — APPROVED** [review]
looks good
